apps/src/gamelab/animationListModule.js

```diff
@@ -0,0 +1,605 @@
+ *
+ *  We need to do a migration if the old style gets loaded.
+ *  See setInitialAnimationMetadata for how this works.
+ */
+import _ from 'lodash';
+import {combineReducers} from 'redux';
+import utils from '../utils';
+import {animations as animationsApi} from '../clientApi';
+import {selectAnimation} from './AnimationTab/animationTabModule';
+import {reportError} from './errorDialogStackModule';
+/* global dashboard */
+
+// Args: {SerializedAnimationList} animationList
+const SET_INITIAL_ANIMATION_LIST = 'SET_INITIAL_ANIMATION_LIST';
+// Args: {AnimationKey} key, {SerializedAnimation} data
+export const ADD_ANIMATION = 'ADD_ANIMATION';
+// Args: {number} index, {AnimationKey} key, {SerializedAnimation} data
+export const ADD_ANIMATION_AT = 'ADD_ANIMATION_AT';
+// Args: {AnimationKey} key, {SerializedAnimation} data
+export const EDIT_ANIMATION = 'EDIT_ANIMATION';
+// Args: {AnimationKey} key
+const INVALIDATE_ANIMATION = 'INVALIDATE_ANIMATION';
+// Args: {AnimationKey} key, {string} name
+const SET_ANIMATION_NAME = 'SET_ANIMATION_NAME';
+// Args: {AnimationKey} key
+const DELETE_ANIMATION = 'DELETE_ANIMATION';
+
+// const SET_ANIMATION_NAME = 'SET_ANIMATION_NAME';
+
+// Args: {AnimationKey} key
+const START_LOADING_FROM_SOURCE = 'START_LOADING_FROM_SOURCE';
+// Args: {AnimationKey} key, {Blob} blob, {String} dataURI. Version?
+const DONE_LOADING_FROM_SOURCE = 'DONE_LOADING_FROM_SgOURCE';
+// Args: {AnimationKey} key, {string} version
+const ON_ANIMATION_SAVED = 'ON_ANIMATION_SAVED';
+
+export default combineReducers({
+  list,
+  data
+});
+
+function list(state, action) {
+  state = state || [];
+  switch (action.type) {
+
+    case SET_INITIAL_ANIMATION_LIST:
+      return action.animationList.list;
+
+    case ADD_ANIMATION:
+      return [].concat(
+          state.slice(0),
+          action.key);
+
+    case ADD_ANIMATION_AT:
+      return [].concat(
+          state.slice(0, action.index),
+          action.key,
+          state.slice(action.index));
+
+    case DELETE_ANIMATION:
+      return state.filter(key => key !== action.key);
+
+    default:
+      return state;
+  }
+}
+
+function data(state, action) {
+  state = state || {};
+  var newState;
+  switch (action.type) {
+
+    case SET_INITIAL_ANIMATION_LIST:
+      return action.animationList.data;
+
+    case ADD_ANIMATION:
+    case ADD_ANIMATION_AT:
+    case EDIT_ANIMATION:
+    case INVALIDATE_ANIMATION:
+    case SET_ANIMATION_NAME:
+    case START_LOADING_FROM_SOURCE:
+    case DONE_LOADING_FROM_SOURCE:
+    case ON_ANIMATION_SAVED:
+      newState = Object.assign({}, state);
+      newState[action.key] = datum(newState[action.key], action);
+      return newState;
+
+    case DELETE_ANIMATION:
+      newState = Object.assign({}, state);
+      delete newState[action.key];
+      return newState;
+
+    default:
+      return state;
+  }
+}
+
+/**
+ * Reducer for a single animation data item.
+ */
+function datum(state, action) {
```

Comment:
I've gone with the following.  It's verbose, but I've been mired in confusion around these and I feel like this will clear it up quite a bit:
```
AnimationList: {
  orderedKeys: [key1, key2],
  propsByKey: {
    key1: {/* AnimationProps */}
    key2: {/* AnimationProps */}
  }
}
```

See the new [gamelab/PropTypes.js](https://github.com/code-dot-org/code-dot-org/blob/6a937299389fe541185b39a13b09e1beba6f4347/apps/src/gamelab/PropTypes.js) for detailed descriptions, typedefs, and React shapes for these.